Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs

```diff
@@ -0,0 +1,223 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Web.Development.Analyzers
+{
+  [DiagnosticAnalyzer (LanguageNames.CSharp)]
+  public class RMWEB0001_WebStringAnalyzer : DiagnosticAnalyzer
+  {
+    public static readonly DiagnosticDescriptor DiagnosticDescriptor = new(
+        "RMWEB0001",
+        "Wrong WebString or PlainTextString usage",
+        "'{0}' should not be used with a '{1}' argument. {2}",
+        "Usage",
+        DiagnosticSeverity.Error,
+        true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create (DiagnosticDescriptor);
+
+    public override void Initialize (AnalysisContext context)
+    {
+      context.EnableConcurrentExecution();
+      context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.Analyze);
+
+      context.RegisterCompilationStartAction (
+          compilationContext =>
+          {
+            var webStringTypeSymbol = compilationContext.Compilation.GetTypeByMetadataName ("Remotion.Web.WebString");
+            var plainTextStringTypeSymbol = compilationContext.Compilation.GetTypeByMetadataName ("Remotion.Web.PlainTextString");
+            var stringBuilderTypeSymbol = compilationContext.Compilation.GetTypeByMetadataName ("System.Text.StringBuilder");
+            var htmlTextWriterSymbol = compilationContext.Compilation.GetTypeByMetadataName ("System.Web.UI.HtmlTextWriter");
+
+            if (webStringTypeSymbol == null
+                || plainTextStringTypeSymbol == null
+                || stringBuilderTypeSymbol == null
+                || htmlTextWriterSymbol == null)
+            {
+              return;
+            }
+
+            var objectSymbol = compilationContext.Compilation.GetTypeByMetadataName ("System.Object");
+            var stringSymbol = compilationContext.Compilation.GetTypeByMetadataName ("System.String");
+            var webStringEnumerable = compilationContext.Compilation.GetTypeByMetadataName ("System.Collections.Generic.IEnumerable`1")!
+                .Construct (webStringTypeSymbol);
+            var plainTextStringEnumerable = compilationContext.Compilation.GetTypeByMetadataName ("System.Collections.Generic.IEnumerable`1")!
+                .Construct (plainTextStringTypeSymbol);
+            var stringBuilderAppendMethodSymbol = GetStringBuilderAppendObjectMethodSymbol (stringBuilderTypeSymbol, objectSymbol);
+            var htmlTextWriterWriteMethodSymbol = GetHtmlTextWriterWriteObjectMethodSymbol (htmlTextWriterSymbol, objectSymbol);
+            var stringJoinMethodSymbols = GetStringJoinMethodSymbols (stringSymbol!);
```

Comment:
Why do we have the null-forgiving here?

Reply:
`GetTypeByMetadataName` returns `INamedTypeSymbol?`, `null` if the type cannot be found. I did not add nullchecks for `object` and `string`, as I expect them to be available. That's why there's the nullforgiving operator here (the logic of `GetStringJoinMethodSymbols` requires the passed symbol to not be `null`). As for the following questions on why some parameters are nullable, `SymbolEqualityComparer.Default.Equals` accepts nullable symbols, that's why the `object` symbol passed in might aswell be `null`` - even if it won't be. It just was a decision between `?` and `!` I made. But I don't really have a strong opinion here.

Reply:
okay, thanks (meaning we have good reasons for the inconsistency). (things you can't see in github)